src/cliffordsynthesis/TargetMetricHandler.cpp

```diff
@@ -0,0 +1,142 @@
+
+    void TargetMetricHandler::makeTargetMetric(const SynthesisData& data, const Configuration& configuration) {
+        bool useMaxSat = configuration.strategy == OptimizationStrategy::UseMinimizer || configuration.strategy == OptimizationStrategy::SplitIter;
+        bool onlyCnot  = configuration.target == TargetMetric::TWO_QUBIT_GATES;
+
+        switch (configuration.target) {
+            case TargetMetric::GATES:
+            case TargetMetric::TWO_QUBIT_GATES:
+                makeGateMetric(data, useMaxSat, onlyCnot);
+                break;
+            case TargetMetric::DEPTH:
+                makeDepthMetric(data, useMaxSat);
+                break;
+            case TargetMetric::FIDELITY:
+                makeFidelityMetric(data, useMaxSat, configuration.architecture);
+                break;
+        }
+    }
+    void TargetMetricHandler::makeGateMetric(const SynthesisData& data, bool useMaxSat, bool onlyCNOT) {
+        logicbase::LogicTerm changes = logicbase::LogicTerm(true);
+        // COST
+        if (useMaxSat) {
+            logicbase::LogicTerm cost = logicbase::LogicTerm(0);
+            for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+                for (unsigned int a = 0; a < data.nqubits; ++a) {
+                    if (!onlyCNOT) {
+                        for (auto gate: Gates::SINGLE_QUBIT_WITHOUT_NOP) {
+                            cost = cost + data.gS[gateStep][Gates::toIndex(gate)][a];
+                        }
+                    }
+                    for (unsigned int b = 0; b <= a; ++b) {
+                        if (a == b) {
+                            continue;
+                        }
+                        cost = cost + data.gTwoQubit[gateStep][a][b] + data.gTwoQubit[gateStep][b][a];
+                    }
+                }
+            }
+            dynamic_cast<logicbase::LogicBlockOptimizer*>(data.lb.get())->minimize(cost);
+        }
+    }
+
+    void TargetMetricHandler::makeDepthMetric(const SynthesisData& data, bool useMaxSat) {
+        using namespace logicbase;
+        // COST
+        if (useMaxSat) {
+            LogicTerm cost = LogicTerm(0);
+            for (unsigned int gateStep = 1; gateStep < data.timesteps + 1; ++gateStep) {
+                LogicTerm anyGate = LogicTerm(true);
+                for (unsigned int a = 0; a < data.nqubits; ++a) {
+                    for (auto gate: Gates::SINGLE_QUBIT_WITHOUT_NOP) {
+                        anyGate = anyGate && !data.gS[gateStep][Gates::toIndex(gate)][a];
+                    }
+                    for (unsigned int b = 0; b <= a; ++b) {
+                        if (a == b) {
+                            continue;
+                        }
+                        anyGate = anyGate && !data.gTwoQubit[gateStep][a][b] && !data.gTwoQubit[gateStep][b][a];
+                    }
+                }
+                cost = cost + LogicTerm::ite(anyGate, LogicTerm(5), LogicTerm(0));
```

Comment:
I might be stupid, but why is the cost increased by `5` here if there is a gate at a particular timestep?
Doesn't this just unnecessarily increase the cost?